### DISCUSSION TITLE
Why does Cap'n Proto not encode a DAG?

### DISCUSSION THREAD

**Discussion by mikehearn** [capnproto/capnproto#2006] [Q&A] [answered]
I didn't see a discussion of why this is the case in the encoding (and am feeling a bit lazy so didn't try to work out the rationale myself). I have a data structure that's best encoded as a DAG, and cpnp uses pointers so would intuitively be able to do this. Under "Amplification attacks" cyclic pointers are already discussed as something the implementation has to guard against anyway, and fully cyclic graphs are a stronger model than just a DAG. Why the limitation to encoding trees?

**Comment by kentonv** [comment, marked as answer]
If aliases of any type were supported, then many of Cap'n Proto's basic operations become much more complicated and inefficient to implement.

For example, you can copy a whole sub-tree from one message to another like:

```
dest.setFoo(source.getFoo());
```

Some use cases use such setters a lot! Since everything is strictly a tree, the implementation is very simple, it just recursively copies. If aliases are allowed, then we have to maintain a map to remember what objects we've seen before, to avoid duplicating things. That's significant overhead, probably requiring memory allocation and lots of map lookups. And probably, we'd have to incur this cost for every copy whether or not it actually contains aliases, because how would we know before we check?

Even if we did all that bookkeeping, there's still a problem. What if I do something like:

```
dest.setFoo(source.getFoo());
dest.setBar(source.getBar());
```

and both "foo" and "bar" turn out to contain aliases to common inner objects? The shared objects will presumably get cloned, unless we actually keep track of the history of copies between `dest` and `source`, which seems unreasonable.

Another point: When you overwrite a pointer in a capnp object, we recursively zero out the no-longer-reachable data. If aliases were allowed, this wouldn't be safe, unless we did some sort of garbage collection. Probably we'd instead just _not_ zero things out, but that's a little bit dangerous as sometimes people assume they can do things like remove a secret string from a message and then forward it on, and this won't leak the secret.

Basically, allowing aliases creates tons of complexity and that complexity seems to affect even the vast majority of use cases that don't need aliases. I decided it was better to make a rule against it.

Instead, the way I recommend representing a graph is to have an array of nodes, and have the nodes refer to each other by index. (I've occasionally even considered adding a way to declare that a field is an index into some parent array, but never had enough of a need to implement it...)

**Reply by mikehearn** [reply]
Thanks for the thorough answer. Yes, I'm doing it with indexes into an array. 

Interestingly, these are all write-side concerns. I guess the read side can handle full graph structures nearly transparently, if you have a writer that doesn't mind doing the work to create them?

**Reply by kentonv** [reply]
I suppose you could say that. The only defense against malicious cycles or aliasing on the read size is the traversal limit. Until that limit kicks in, there's no detection of aliasing, so readers will accept arbitrary graphs in a sense.

Of course, if the reader tries to copy the message into their own builder, they'll have trouble.

Really, the biggest problem is in proxies. Proxies need to copy messages from incoming requests to outgoing requests all the time. Cap'n Proto RPC makes it very easy to set up proxying, but if you have messages passing through that are DAGs, they are going to be exploded into trees implicitly, only raising an error if the message ends up crossing the (by default) 64MB traversal limit.
